Github 로그인 구현

EarlyHail opened the pull request:
### 구현의도
- Github 로그인 구현
  - github oauth의 callback api
     1. github의 인증 서버로부터 accessToken을 받아온다.
     2. github의 user 데이터 서버에 접근해 유저 정보를 받아온다.

- Cookie에 저장된 jwt 토큰 verify 구현
  `routers/index.ts`에 미들웨어로 처리, auth를 조건문으로 필터링하도록 구현
  verify 결과로 나온 사용자의 ObjectId (`_id` 필드)를 `ctx.state.user={ _id: ObjectId }`형태로 전달

- Mongoose User 모델 구현
   필드는 uid (github의 uid)와 nickname (github 별명), email (github에서 제공될 시)로 구성

### 기능 흐름도, 클래스 다이어그램(선택사항)
- 

### 사용된 기술(선택사항)
- 

### 리뷰 & 논의사항 & 궁금한점(선택사항)
-

EarlyHail committed:
[feat] github 로그인 기능 추가

- Oauth를 github api 작성

- github oauth의 callback api 작성
   1. github의 인증 서버로부터 accessToken을 받아온다.
   2. github의 user 데이터 서버에 접근해 유저 정보를 받아온다.

- fetch를 위한 node-fetch 설치

EarlyHail committed:
[chore] .env.template 수정

- github client id와 client secret 추가로 인한 .env 템플릿 수정

EarlyHail committed:
[feat] User model 추가

- user model 추가
  github의 uid, 별명, email을 필드로 가지도록 생성
  email은 github 사용자의 게정 설정에 따라 존재 유무가 다르므로 nullable이다.

EarlyHail committed:
[feat] github 로그인 유저 정보 DB 저장

- 기존 github 로그인 로직을 Service 계층으로 분리

- User data를 받아온 뒤 DB에 저장
  - UserType을 만들어 validation 후, findOneAndUpdate를 이용하여
    데이터가 없으면 inert, 데이터가 있을시 update를 하도록 기능 구현

EarlyHail committed:
[feat] jwt 토큰 발행 기능 추가

- 로그인 성공 시, jwt 토큰을 발행하는 기능 추가

- jsonwebtoken을 이용하여 token 발행, cookie에 저장 후 프론트 서버로 리다이렉팅

- .env.template에 JWT Secret 추가

EarlyHail committed:
[fix] expiration 시간 변경, 시간 상수화, warning 해결

- expiration 시간 변경
   1일 -> 3시간으로 변경

- expiration 값 변수화
  `githubCallback` controller에서 expiration 값을 설정

- mongoose 설정에 `useFindAndModify: false` 설정 추가
  `findOneAndUpdate` 실행 시 warning 해결

EarlyHail committed:
[feat] jwt 토큰 verify 기능 추가

- 토큰 확인 middleware를 `routes/index.ts`에 설정

- middleware에서는 api가 auth 관련이면, 다음 라우터로 넘긴다.
   auth 라우터가 아니면, cookies에서 토큰을 뽑아와 확인한다.
  토큰을 verify한 뒤, `state.user`에 디코딩된 토큰의  `_id`를 담아 다음 라우터에게 전달한다.

EarlyHail committed:
[refactor] OAuth front redirect url 환경변수로 사용

- OAuth의 frontend redirect url을 환경 변수로 사용

EarlyHail committed:
[refactor] .env 파일을 가져올 때 타입 단언 사용

- 기존에 string | undefined로 사용했던 .env를 타입단언으로 수정

EarlyHail committed:
[fix] cookie 저장 옵션 httpOnly false로 변경

- cookie 저장 옵션을 httpOnly로 변경
  true일 경우 Admin의 브라우저에서 토큰 값을 읽지 못함
  (React가 로그인 판별 불가)

- 로그인 방식에 대해 추후 논의 필요

juyoungpark718 commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -0,0 +1,34 @@
+import fetch from 'node-fetch';
+
+const clientId = process.env.GITHUB_OAUTH_CLIENT_ID;
+const clientSecret = process.env.GITHUB_OAUTH_CLIENT_SECRET;
```
```suggestion
const CLIENT_ID = process.env.GITHUB_OAUTH_CLIENT_ID;
const CLIENT_SECRET = process.env.GITHUB_OAUTH_CLIENT_SECRET;
```

변수 선언을 대문자로 바꾸는거 어떻게 생각하세요??

juyoungpark718 commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -0,0 +1,34 @@
+  // ctx.body = { profile: profile.login };
+  ctx.redirect('http://localhost:3001');
+  await next();
+};
```
query에 code가 없는 경우가 있을까요??

juyoungpark718 commented on src/models/User.ts:
```diff
@@ -0,0 +1,26 @@
+import { Schema, Document, model, Model } from 'mongoose';
+
+export interface UserType {
+  uid: number;
+  nickname: string;
+  email: string | undefined;
+}
```
github 소셜로그인밖에 없는데, profile까지 받아오면 좋을 것 같은데 어떻게 생각하세요??

juyoungpark718 commented on src/routes/auth/services/githubUtil.ts:
```diff
@@ -0,0 +1,48 @@
+    },
+  });
+  const profile: any = await profileResponse.json();
+  const newUser: UserDocument | null = await insertUser(profile);
```
any타입보다는 profileResponse의 값을 확인해보고 type을 정의해주는 것에 대해서는 어떻게 생각하세요??

juyoungpark718 commented on src/utils/authMiddleware.ts:
```diff
@@ -0,0 +1,20 @@
+    const jwtToken = ctx.cookies.get('token');
+    if (!jwtToken) ctx.throw(401, 'login first');
+    const userId = checkToken(jwtToken);
+    ctx.state.user = { _id: userId };
```
'login first'같은 에러 메세지에 대한 처리는 다같이 상의해서 정리해보는 것도 나쁘지 않다는 생각이 듭니다.

그리고 checkToken에서 undefined값을 받았을때에 대한 오류 처리를 backend, frontend에서 하는지에 대해서도 같이 생각해보면 좋을 것 같습니다.

saeeng commented on src/models/User.ts:
```diff
@@ -0,0 +1,26 @@
+const userSchema = new Schema({
+  uid: { type: Number, required: true },
+  nickname: { type: String, required: true },
+  email: { type: String },
```
email은 required 할 필요가 없을까요?? 깃헙자체가 email이 필수라 없는 경우는 없을것 같아요!

EarlyHail replied:
저번에 github OAuth할 때 Email정보를 보내줄껀지 사용자가 확인해야한다는 이야기가 기억이 나서 일단 지워놨습니다.

제가 로그인 했을 때도 넘어오는 데이터에 email이 null이더라구요....

saeeng replied:
github Oauth 요청시에 scope : email  이런식으로 보내면 권한 요청 부분에서 email까지 같이 승인하도록 할거에요!
```
    const query = qs.stringify({
        scope: 'user:email',
    });
```

EarlyHail replied:
![image](https://user-images.githubusercontent.com/55068119/100755368-74a4d980-342f-11eb-9d10-d6abeb3087bb.png)
잘 되네요! 그런데 위에 말했던것처럼 사용자가 Public Email을 등록하지 않으면 null로 나옵니다.

![image](https://user-images.githubusercontent.com/55068119/100755559-a74ed200-342f-11eb-92fb-afbf49c04b2a.png)
이렇게 Public email을 등록해줘야 받을 수 있네요.... 일단 email 권한 요청 부분 수정해서 올리겠습니다

saeeng replied:
으흠 그렇군요... 회원가입시에 받는 이메일과는 또 다른 이메일을 전송해주나보네요...

EarlyHail committed:
[fix] OAuth scope 수정

- github OAuth scope 수정
   사용자의 email을 받아올 수 있도록 수정
   사용자가 public email을 등록하지 않으면 null이므로 모델은 일단 변경하지 않음

EarlyHail committed:
[fix] 쿠키에 nickname 추가

- 사용자 별명이 일단 Admin에 필요하므로 쿠키 추가

junsushin-dev commented on .env.template:
```diff
@@ -1 +1,4 @@
+MONGO_URI=
+
+GITHUB_OAUTH_CLIENT_ID=
+GITHUB_OAUTH_CLIENT_SECRET=
```
process.env 쪽 추가되고 바로 템플릿에 반영해주신 점 좋네요! 😋

junsushin-dev commented on src/routes/auth/services/githubUtil.ts:
```diff
@@ -0,0 +1,48 @@
+  const result = await User.findOneAndUpdate(
+    { uid: newUser.uid },
+    { $setOnInsert: newUser },
+    {
+      upsert: true,
+    },
```
upsert가 update + insert의 약자인가 보네요! 하나 배워갑니다. 🙂

junsushin-dev commented on src/routes/auth/services/githubUtil.ts:
```diff
@@ -1,6 +1,9 @@
 import User, { UserType, UserDocument } from '../../../models/User';
 
+require('dotenv').config();
+
```
저희 app의 엔트리포인트 최상단에 `require('dotenv').config();` 가 이미 있어서 없어도 될 것 같습니다! 😗

junsushin-dev commented on src/routes/auth/services/githubUtil.ts:
```diff
@@ -47,13 +47,14 @@ const processGithubOAuth = async (code: string): Promise<UserDocument | null> =>
-    const accessToken = jwt.sign({ userId }, jwtSecret, { expiresIn: '1d' });
-    return accessToken;
+    const jwtToken = jwt.sign({ _id: userId }, jwtSecret, { expiresIn: tokenExpiration });
+    return jwtToken;
```
OAuth access token과 혼동될 가능성을 없애주는 좋은 이름 변경이네요! 👍

junsushin-dev commented on src/utils/checkToken.ts:
```diff
@@ -0,0 +1,14 @@
+import jwt from 'jsonwebtoken';
+
+require('dotenv').config();
+
```
이쪽도 비슷하게 entry point 최상단(app.ts)에서 dotenv를 한번 돌려주기 주기 때문에 빼도 될 것 같습니다.

junsushin-dev commented on src/utils/checkToken.ts:
```diff
@@ -0,0 +1,14 @@
+    // eslint-disable-next-line no-underscore-dangle
+    return userId._id;
+  }
+  return undefined;
```
```suggestion
  const jwtSecret: string = process.env.JWT_SECRET as string;
  const userId: any = jwt.verify(jwtToken, jwtSecret);
  // eslint-disable-next-line no-underscore-dangle
  return userId._id;
```

process.env 가 없는 경우는 서버가 구동하는 도중에 처리가 필요한 예외는 아닌 것 같아서, 이런 식으로 바꾸어서 lint 에러를 피해가는 것이 더 간결할 수 있을 것 같네요 ☺

junsushin-dev commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -2,6 +2,8 @@ import { Context, Next } from 'koa';
 import { UserDocument } from '../../../models/User';
 import { processGithubOAuth, getToken } from '../services/githubUtil';
 
+require('dotenv').config();
```
이쪽도 빼주어도 무방할 것 같습니다!

junsushin-dev commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -12,6 +14,6 @@ export default async (ctx: Context, next: Next): Promise<void> => {
     ctx.cookies.set('token', jwtToken, { httpOnly: true, maxAge: tokenExpiration });
   }
-  ctx.redirect('http://localhost:3001');
+  ctx.redirect(process.env.ADMIN_MAIN_URL as string);
```
코멘트를 달았었는데, 이미 바꾸셨군요! 변수로 바꾸신 것 아주 좋은 것 같습니다. 👍 
굳이 .env에 숨겨야 하는 정보는 아니라서 그냥 변수로 두어도 될까 싶었는데, development / production 환경에 따라서 코드를 바꾸어주지 않아도 되는 장점이 있을 것 같네요.

junsushin-dev commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -9,10 +9,15 @@ const tokenExpiration: number = 3 * HOUR;
+     * @TODO
+     * http only option
+     */
   }
```
newUser가 UserDocument 타입으로 보장된다면, getToken 함수 반환값을 string으로 고정하고 jwtToken의 타입에서도 undefined는 빼주어도 될 것 같습니다! 😁 (processGithubOAuth 함수 쪽에서 DB 요청 비동기 에러가 나면 에러핸들러로 보내주어서 newUser 값이 보장된다면 더욱 좋을 것 같네요)
```suggestion
  const newUser: UserDocument = (await processGithubOAuth(accessCode)) as UserDocument;
  const jwtToken: string = getToken(newUser, tokenExpiration);
  ctx.cookies.set('nickname', newUser.nickname, { httpOnly: false, maxAge: tokenExpiration });
  ctx.cookies.set('token', jwtToken, { httpOnly: false, maxAge: tokenExpiration });
  /**
   * @TODO
   * http only option
   */
```

junsushin-dev reviewed (commented):
제가 앞에서 코멘트를 달았던 부분이 뒤에 보니까 고쳐놓으신 부분이 많아서 사람 생각이 다 똑같구나 생각했네요. 이외 코멘트 몇 개 달았는데 한번 확인 부탁드려요! 😁

junsushin-dev commented on src/routes/auth/controllers/githubCallback.ts:
```diff
@@ -1,34 +1,11 @@
 
-  // ctx.body = { profile: profile.login };
   ctx.redirect('http://localhost:3001');
   await next();
```
이미 나중 commit에서 바꾸신 부분이라 resolve 처리합니다.

EarlyHail committed:
[refactor] 불필요 .env 코드 삭제, 타입 단언문 사용

- 불필요한 .env require문 삭제

- jwtToken 토큰에 타입 단언 사용하여 불필요한 undefined 타입 코드 제거

EarlyHail committed:
[refactor] 불필요 .env 상수화 제거, 타입단언 사용

- 불필요 .env 상수화 제거
  한번만 사용하는 .env를 전역으로 빼놓은 코드 삭제

- 타입 단언 사용으로 불필요 undefined 타입 선언 제거

juyoungpark718 reviewed (approved):
정말 고생하셨습니다. OAuth 너무 쉽지 않았을텐데...!! 고생하셨습니다 !!

junsushin-dev reviewed (approved):
고생하셨습니다 호진님! 저와 주영님의 리뷰가 주렁주렁 달렸는데 잘 대응해주셨네요 🤣

EarlyHail committed:
[fix] Github 로그인 Cancel시 처리

- Github 로그인 Cancel 버튼을 누를 시 처리
   Cancel을 누를 시 오는 Error를 체크 후 메인 페이지로 리다이렉팅

EarlyHail committed:
[fix] github 권한 에러 수정, profile 타입 수정

- github 권한 에러 수정
  github 권한 수락 안할 시 error로 처리해야하는데 code로 처리한 로직 수정

- profile 타입 수정
  기존 any 타입에서 필요한 타입들로 interface를 선언하여 any 삭제

EarlyHail committed:
Merge branch 'develop' into feature/github-oauth

saeeng reviewed (approved):
수고하셨습니다 👍
